TrackletAlgorithm/TrackBuilder.h

```diff
@@ -340,24 +359,137 @@ void TrackBuilder(
     }
     nTracks += (track.getTrackValid() ? 1 : 0);
 
-    // Update the circular buffer indices and read a new element from each of
-    // the input full-match memories.
-    barrel_circular_buffer_update : for (short j = 0; j < NFMBarrel; j++) {
-      barrel_read_index[j] += (barrel_valid[j] ? 1 : 0);
-      const ap_uint<kNBitsTBBuffer> barrel_next_write_index = barrel_write_index[j] + 1;
-      const ap_uint<1> barrel_not_full = (barrel_next_write_index != barrel_read_index[j]);
-      getFM<BARREL>(bx, barrelFullMatches[j], barrel_mem_index[j], barrel_fm[j][barrel_write_index[j]]);
-      barrel_mem_index[j] += ((empty || barrel_not_full) ? 1 : 0);
-      barrel_write_index[j] += ((empty || barrel_not_full) ? 1 : 0);
-    }
+    // Do the next set of pairwise comparisons in the merge trees.
+    barrel_merger : for (unsigned short j = 0; j < NBarrelStubs; j++) {
+
+      bool read_1, read_2;
+
+      merger_L_top[j].next(merger_L_b1[j].peek(), merger_L_b1[j].valid(), read_1,
+                            merger_L_b2[j].peek(), merger_L_b2[j].valid(), read_2,
+                            smallest[j]);
+
+      // We use the second layer of the merge tree only when there are eight FM
+      // memories per layer, which should only occur for the first layer.
```

Comment:
I wonder if there is a simpler way to implement this?

Reply:
@aryd I wonder the same thing, but having stared at it for so long, I don't immediately see a way. Any ideas?

Reply:
Not really...